pkg/objstore/swift/swift.go

```diff
@@ -0,0 +1,267 @@
+// Package swift implements common object storage abstractions against OpenStack swift APIs.
+package swift
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"math/rand"
+	"os"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/improbable-eng/thanos/pkg/objstore"
+
+	"github.com/go-kit/kit/log"
+	"github.com/gophercloud/gophercloud"
+	"github.com/gophercloud/gophercloud/openstack"
+	"github.com/gophercloud/gophercloud/openstack/objectstorage/v1/containers"
+	"github.com/gophercloud/gophercloud/openstack/objectstorage/v1/objects"
+	"github.com/gophercloud/gophercloud/pagination"
+	"github.com/pkg/errors"
+	"gopkg.in/yaml.v2"
+)
+
+// DirDelim is the delimiter used to model a directory structure in an object store bucket.
+const DirDelim = "/"
+
+type swiftConfig struct {
+	AuthUrl       string `yaml:"auth_url"`
+	Username      string `yaml:"username,omitempty"`
+	UserId        string `yaml:"user_id,omitempty"`
+	Password      string `yaml:"password"`
+	DomainId      string `yaml:"domain_id,omitempty"`
+	DomainName    string `yaml:"domain_name,omitempty"`
+	TenantID      string `yaml:"tenant_id,omitempty"`
+	TenantName    string `yaml:"tenant_name,omitempty"`
+	RegionName    string `yaml:"region_name,omitempty"`
+	ContainerName string `yaml:"container_name"`
+}
+
+type Container struct {
+	logger log.Logger
+	client *gophercloud.ServiceClient
+	name   string
+}
+
+func NewContainer(logger log.Logger, conf []byte) (*Container, error) {
+	var sc swiftConfig
+	if err := yaml.Unmarshal(conf, &sc); err != nil {
+		return nil, err
+	}
+
+	authOpts := gophercloud.AuthOptions{
+		IdentityEndpoint: sc.AuthUrl,
+		Username:         sc.Username,
+		UserID:           sc.UserId,
+		Password:         sc.Password,
+		DomainID:         sc.DomainId,
+		DomainName:       sc.DomainName,
+		TenantID:         sc.TenantID,
+		TenantName:       sc.TenantName,
+
+		// Allow Gophercloud to re-authenticate automatically.
+		AllowReauth: true,
+	}
+
+	provider, err := openstack.AuthenticatedClient(authOpts)
+	if err != nil {
+		return nil, err
+	}
+
+	client, err := openstack.NewObjectStorageV1(provider, gophercloud.EndpointOpts{
+		Region: sc.RegionName,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	c := &Container{
```

Comment:
You could potentially inline this, but not a blocker.

Reply:
done